src/dmd/dmodule.d

```diff
@@ -653,188 +653,199 @@ extern (C++) final class Module : Package
-                                    fatal();
-                                }
-                                u2 = le ? Port.readwordLE(pu) : Port.readwordBE(pu);
-                                if (u2 < 0xDC00 || u2 > 0xDFFF)
-                                {
-                                    error("surrogate UTF-16 low value %04x out of range", u2);
-                                    fatal();
-                                }
-                                u = (u - 0xD7C0) << 10;
-                                u |= (u2 - 0xDC00);
-                            }
-                            else if (u >= 0xDC00 && u <= 0xDFFF)
+                            uint u2;
+                            if (++pu > pumax)
                             {
-                                error("unpaired surrogate UTF-16 value %04x", u);
+                                error("surrogate UTF-16 high value %04x at end of file", u);
                                 fatal();
                             }
-                            else if (u == 0xFFFE || u == 0xFFFF)
+                            u2 = readNext(pu);
+                            if (u2 < 0xDC00 || u2 > 0xDFFF)
                             {
-                                error("illegal UTF-16 value %04x", u);
+                                error("surrogate UTF-16 low value %04x out of range", u2);
                                 fatal();
                             }
-                            dbuf.writeUTF8(u);
+                            u = (u - 0xD7C0) << 10;
+                            u |= (u2 - 0xDC00);
                         }
-                        else
-                            dbuf.writeByte(u);
+                        else if (u >= 0xDC00 && u <= 0xDFFF)
+                        {
+                            error("unpaired surrogate UTF-16 value %04x", u);
+                            fatal();
+                        }
+                        else if (u == 0xFFFE || u == 0xFFFF)
+                        {
+                            error("illegal UTF-16 value %04x", u);
+                            fatal();
+                        }
+                        dbuf.writeUTF8(u);
                     }
-                    dbuf.writeByte(0); // add 0 as sentinel for scanner
-                    buflen = dbuf.offset - 1; // don't include sentinel in count
-                    buf = dbuf.extractData();
                 }
+                else
+                    dbuf.writeByte(u);
+            }
+            dbuf.writeByte(0); //add a terminating null byte
+            return dbuf.extractSlice();
+        }
+
+        const(char)* srcname = srcfile.name.toChars();
+        //printf("Module::parse(srcname = '%s')\n", srcname);
+        isPackageFile = (strcmp(srcfile.name.name(), "package.d") == 0 ||
+                         strcmp(srcfile.name.name(), "package.di") == 0);
```

Comment:
can you use strings here?

Reply:
That will require a few more changes.  I'll take a look at that next.

Reply:
OK, Thanks.